Unit testing and audio playback

LukaszTheLittle opened the pull request:
Added functionality of playing all of the .wav files. 
Wrote unit tests for `SoundViewModel` integration with `BeatBox`.

LukaszTheLittle committed:
created `SoundPool`

LukaszTheLittle committed:
Loaded sounds into SoundPool

LukaszTheLittle committed:
Loaded up all sounds

LukaszTheLittle committed:
Commit

LukaszTheLittle committed:
Added function `play` to play sounds back

LukaszTheLittle committed:
Updated dependencies

LukaszTheLittle committed:
Added Mockito dependencies

LukaszTheLittle committed:
Created test class `SoundViewModelTest.kt`

LukaszTheLittle committed:
Added `JUnit4` for testing

LukaszTheLittle committed:
Removed unnecessary `testImplementation`

LukaszTheLittle committed:
Created a `SoundViewModel` test subject

LukaszTheLittle committed:
Created function to test the title property

LukaszTheLittle committed:
Changed deprecated `assertThat` to `assertSame`

LukaszTheLittle committed:
Removed unnecessary ``is``

LukaszTheLittle committed:
Wrote test for `onButtonClicked()`

LukaszTheLittle committed:
Created `onButtonClicked()`

LukaszTheLittle committed:
Created a mock BeatBox

LukaszTheLittle committed:
Verify that `BeatBox.play(sound)` is called

LukaszTheLittle committed:
Provided `BeatBox` to `SoundViewModel`

LukaszTheLittle committed:
Fixed error in `SoundHolder`

LukaszTheLittle committed:
Provided the mock `BeatBox`

LukaszTheLittle committed:
Implemented `onButtonClicked()`

LukaszTheLittle committed:
Updated release of mockito

LukaszTheLittle committed:
Hooked up buttons

LukaszTheLittle committed:
Released `SoundPool`

LukaszTheLittle committed:
Released `BeatBox`

Pantkowsky commented on app/src/main/java/com/bignerdranch/android/beatbox/MainActivity.kt:
```diff
@@ -27,11 +27,16 @@ class MainActivity : AppCompatActivity() {
 
         init {
-            binding.viewModel = SoundViewModel()
+            binding.viewModel = SoundViewModel(beatBox)
```
is `beatBox` here a global variable of `MainActivity` ? If yes, then it would be better if you passed it as an argument to the `SoundHolder` instead of referencing it as a global variable and using it as an arg to `SoundViewModel`. When creating `private inner class` (but generally, when passing arguments around), consider how much work will you have when you need to refactor it, by for example moving it to separate file. If `SoundHolder` accepted `beatBox` as an argument, you would have much less refactoring then, and the code would be cleaner. Remember the topic of _side effects_ ?

LukaszTheLittle replied:
You have a good point. I removed global variable as you advised and passed it as an argument wherever it was needed.

LukaszTheLittle committed:
Removed global variable `private lateinit var beatBox`
Changed argument in `SoundAdapter`
Changed `beatBox` method to `BeatBox(assets)` in `onDestroy()`
Added argument `beatBox` in `private inner class SoundHolder`

LukaszTheLittle committed:
Added `private val soundAdapter`
Changed adapter in `binding.recyclerView.apply`
Passed in `onDestroy()` refactored method for beatBox release function
Changed constructor in `SoundAdapter` class
Refactored arguments in `SoundHolder`
Refactored `val sound` in `onBindViewHolder` method
Refactored `getItemCount` method
Created `beatBoxRelease` method

LukaszTheLittle committed:
Moved `SoundHolder` class to another file

LukaszTheLittle committed:
Deleted unused import directives

LukaszTheLittle committed:
Moved `SoundAdapter` class to another file
